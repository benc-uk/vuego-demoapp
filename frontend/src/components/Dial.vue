<!--
Dial/gauge component, wrapper around gauge.js library
-----------------------------------------------------
Ben C, April 2018, Updated for Vue3 2021
-->

<template>
  <div class="card border-primary">
    <canvas ref="can">{{ value }}</canvas>
    <h5>{{ title }}: {{ valComputed }}<span v-if="percentage">%</span></h5>
  </div>
</template>

<script>
import { Gauge } from 'gaugeJS'

const staticZones = [
  { strokeStyle: '#30B32D', min: 0, max: 70 },
  { strokeStyle: '#FFDD00', min: 70, max: 90 },
  { strokeStyle: '#F03E3E', min: 90, max: 100 }
]

const opts = {
  angle: -0.15,
  lineWidth: 0.3,
  radiusScale: 1,
  pointer: {
    length: 0.5,
    strokeWidth: 0.045,
    color: '#4ce1ff'
  },
  colorStart: '#2063b9',
  limitMax: false,
  limitMin: false,
  strokeColor: '#ddd',
  generateGradient: true,
  highDpiSupport: true,
  fontSize: 40
}

export default {
  name: 'Dial',

  props: {
    value: {
      default: 0,
      type: Number
    },
    title: {
      default: 'No Label',
      type: String
    },
    percentage: {
      default: true,
      type: Boolean
    }
  },

  data: function () {
    return {
      gauge: null,
      opts: opts
    }
  },

  computed: {
    valComputed: function () {
      const rounded = Math.round(this.value * Math.pow(10, 2)) / Math.pow(10, 2)
      if (this.value && this.gauge) {
        this.gauge.set(rounded)
      }
      return rounded
    }
  },

  mounted: function () {
    const target = this.$refs.can

    if (this.percentage) {
      this.opts.staticZones = staticZones
    } else {
      this.opts.staticZones = null
    }
    this.opts.limitMax = this.percentage
    this.gauge = new Gauge(target).setOptions(this.opts)
    this.gauge.maxValue = this.percentage ? 100 : 1
    this.gauge.setMinValue(0)
    this.gauge.set(this.value)
  }
}
</script>

<style scoped>
canvas {
  width: 100%;
  height: 100%;
}

h5 {
  text-align: center;
  color: var(--bs-cyan);
}

div {
  border: 3px solid #ccc;
  border-radius: 6px;
  margin: 5px;
  background-color: rgba(0, 0, 0, 0.4);
}
</style>
