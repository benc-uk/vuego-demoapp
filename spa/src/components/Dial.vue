<!--
Dial/gauge component, wrapper around gauge.js library
-----------------------------------------------------
Ben C, April 2018
-->

<template>
  <div>
    <canvas ref="can">{{ value }}</canvas>
    <h5>{{ title }}: {{ valComputed }}<span v-if="percentage">%</span></h5>
  </div>
</template>

<script>
import { Gauge } from '../js/gauge.min.js'

let staticZones = [
  { strokeStyle: '#30B32D', min: 0, max: 70 },
  { strokeStyle: '#FFDD00', min: 70, max: 90 },
  { strokeStyle: '#F03E3E', min: 90, max: 100 }
]

let opts = {
  angle: -0.15,
  lineWidth: 0.3,
  radiusScale: 1,
  pointer: {
    length: 0.5,
    strokeWidth: 0.045,
    color: '#444'
  },
  colorStart: '#2063b9',
  limitMax: false,
  limitMin: false,
  strokeColor: '#ddd',
  generateGradient: true,
  highDpiSupport: true,
  fontSize: 40
}

export default {
  name: 'Dial',

  props: {
    value: {
      default: 0,
      type: Number
    },
    title: {
      default: 'No Label',
      type: String
    },
    percentage: {
      default: true,
      type: Boolean
    }
  },

  data: function() {
    return {
      gauge: null,
      opts: opts
    }
  },

  computed: {
    valComputed: function() {
      let rounded = Math.round(this.value * Math.pow(10, 2)) / Math.pow(10, 2)
      if (this.value && this.gauge) {
        this.gauge.set(rounded)
      }
      return rounded
    }
  },

  mounted: function() {
    let target = this.$refs.can

    if (this.percentage) {
      this.opts.staticZones = staticZones
    } else {
      this.opts.staticZones = null
    }
    this.opts.limitMax = this.percentage
    this.gauge = new Gauge(target).setOptions(this.opts)
    this.gauge.maxValue = this.percentage ? 100 : 1
    this.gauge.setMinValue(0)
    this.gauge.set(this.value)
  }
}
</script>

<style scoped>
canvas {
  width: 100%;
  height: 100%;
}
h5 {
  text-align: center;
}
div {
  border: 3px solid #ccc;
  border-radius: 6px;
  margin: 5px;
  background-color: white;
}
</style>
