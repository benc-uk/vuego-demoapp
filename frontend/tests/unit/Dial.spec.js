import { mount } from '@vue/test-utils'
import Dial from '@/components/Dial.vue'

describe('Dial.vue', () => {
  it('renders a dial', async () => {
    const wrapper = mount(Dial, {
      propsData: {
        value: 22,
        title: 'Test Dial'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
