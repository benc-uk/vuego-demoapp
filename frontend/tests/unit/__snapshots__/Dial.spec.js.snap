// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dial.vue renders a dial 1`] = `
<div class="card border-primary"><canvas height="0" width="0">22</canvas>
  <h5>Test Dial: 22<span>%</span></h5>
</div>
`;
